pymc3/backends/arviz.py

```diff
@@ -0,0 +1,606 @@
+
+        self.density_dist_obs = density_dist_obs
+        self.observations, self.multi_observations = self.find_observations()
+
+    def find_observations(self) -> Tuple[Optional[Dict[str, Var]], Optional[Dict[str, Var]]]:
+        """If there are observations available, return them as a dictionary."""
+        if self.model is None:
+            return (None, None)
+        observations = {}
+        multi_observations = {}
+        for obs in self.model.observed_RVs:
+            if hasattr(obs, "observations"):
+                aux_obs = obs.observations
+                observations[obs.name] = (
+                    aux_obs.get_value() if hasattr(aux_obs, "get_value") else aux_obs
+                )
+            elif hasattr(obs, "data") and self.density_dist_obs:
+                for key, val in obs.data.items():
+                    aux_obs = val.eval() if hasattr(val, "eval") else val
+                    multi_observations[key] = (
+                        aux_obs.get_value() if hasattr(aux_obs, "get_value") else aux_obs
+                    )
+        return observations, multi_observations
+
+    def split_trace(self) -> Tuple[Union[None, "MultiTrace"], Union[None, "MultiTrace"]]:
+        """Split MultiTrace object into posterior and warmup.
+
+        Returns
+        -------
+        trace_posterior: MultiTrace or None
+            The slice of the trace corresponding to the posterior. If the posterior
+            trace is empty, None is returned
+        trace_warmup: MultiTrace or None
+            The slice of the trace corresponding to the warmup. If the warmup trace is
+            empty or ``save_warmup=False``, None is returned
+        """
+        trace_posterior = None
+        trace_warmup = None
+        if self.save_warmup and self.ntune > 0:
+            trace_warmup = self.trace[: self.ntune]
+        if self.ndraws > 0:
+            trace_posterior = self.trace[self.ntune :]
+        return trace_posterior, trace_warmup
+
+    def log_likelihood_vals_point(self, point, var, log_like_fun):
+        """Compute log likelihood for each observed point."""
+        log_like_val = utils.one_de(log_like_fun(point))
+        if var.missing_values:
+            mask = var.observations.mask
+            if np.ndim(mask) > np.ndim(log_like_val):
+                mask = np.any(mask, axis=-1)
+            log_like_val = np.where(mask, np.nan, log_like_val)
+        return log_like_val
+
+    def _extract_log_likelihood(self, trace):
+        """Compute log likelihood of each observation."""
+        if self.trace is None:
+            return None
+        if self.model is None:
+            return None
+
+        if self.log_likelihood is True:
+            cached = [(var, var.logp_elemwise) for var in self.model.observed_RVs]
+        else:
+            cached = [
+                (var, var.logp_elemwise)
+                for var in self.model.observed_RVs
+                if var.name in self.log_likelihood
```

Comment:
This will need to be updated, because the `var.log*` methods are gone now.  Instead, you can use something like `logpt(var)`.

Reply:
I want all of `log_likelihood_vals_point` and `_extract_log_likelihood` to go and instead evaluate the likelihood in a vectorized fashion, looping over observed variables at most. I am reading the developer guide and trying to see how that would be done.

Reply:
I have given this a try now, and I am not understanding what is happening nor where to find the relevant docs between pymc3 docstrings and aesara. Say I have this model:

```
with pm.Model() as model:
    m = pm.Normal("m", 0.0, 1.0)
    a = pm.Normal("a", mu=m, sigma=1, observed=0.0)
```
I then undestood I have to use `logpt` on `a` to get its pointwise log likelihood: `log_lik = pm.distributions.logpt(a)` and the observed values should be pulled from `a`. I then try to compile a function to calculate log_lik (for now with manual scalar input for m, but ideally with a numpy array containing the whole trace). However, I can't seem to do that:

```
aesara.function([], log_lik)
```

errors out with:

```
...
    raise MissingInputError(error_msg, variable=var)
aesara.graph.fg.MissingInputError: Input 0 (m) of the graph (indices start from 0),
used to compute InplaceDimShuffle{x}(m), was not provided and not given a value. Use
 the Aesara flag exception_verbosity='high', for more information on this error.

Backtrace when that variable is created:

  File "<stdin>", line 2, in <module>
  File "/home/oriol/miniconda3/envs/arviz/lib/python3.8/site-packages/pymc3/distribu
tions/distribution.py", line 96, in __new__
    rv_out = cls.dist(*args, rng=rng, **kwargs)
  File "/home/oriol/miniconda3/envs/arviz/lib/python3.8/site-packages/pymc3/distribu
tions/continuous.py", line 477, in dist
    return super().dist([mu, sigma], **kwargs)
  File "/home/oriol/miniconda3/envs/arviz/lib/python3.8/site-packages/pymc3/distribu
tions/distribution.py", line 105, in dist
    rv_var = cls.rv_op(*dist_params, **kwargs)
```

and 

```
aesara.function([m], log_lik)
```

errors with:

```
...
    raise UnusedInputError(msg % (inputs.index(i), i.variable, err_msg))
aesara.compile.function.types.UnusedInputError: aesara.function was asked to create
a function computing outputs given certain inputs, but the provided input variable a
t index 0 is not part of the computational graph needed to compute the outputs: m.
To make this error into a warning, you can pass the parameter on_unused_input='warn'
 to aesara.function. To disable it completely, use on_unused_input='ignore'.

```

Reply:
Looks like you need to use `m.tag.value_var`.

Reply:
It's rather confusing, because the display names for these two variables (i.e. `m` and `m.tag.value_var`) are exactly the same.  I thought about changing the name of the latter to something like `"{rv_name}_value"` or `"{rv_name}_lik"`, but I couldn't decide.

Anyway, it's very important to understand that there are now **two** distinct, but complementary variables in the model: `TensorVariable`s that represent random variables, and `TensorVariable`s that represent concrete values of those random variables.  This idea is directly related to the `observed` parameter, because it's essentially a concrete instance of the latter type of variable.

More specifically, when one (sloppily) writes `P(Y = y)` in "math"-like notation, `Y` is the random variable and `y` is an "instance" of `Y`, but they are two distinct variables with different properties/constraints.  The latter variable (i.e. the "value" variable), `y`, is the input argument used in density functions, e.g. `f(y) = exp(- y**2 * ...) * ...`; this is why the "value" variable is the only one used in the log-likelihood graphs.

These "value" variables are conveniently stored within their corresponding random variable's `tag` (i.e. a `Variable`'s arbitrary "storage" space) under the name `value_var`.